ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala

```diff
@@ -809,43 +825,106 @@ private[backend] trait CommonStorageBackend[DB_BATCH] extends StorageBackend[DB_
-    List(
+    val conditionalImmediateDivulgencePruning =
+      pruneImmediateDivulgence(pruneUpToInclusive, pruneAllDivulgedContracts)
+
+    val consumingExercisesPruning =
       SQL"""
-          -- Divulgence events (only for contracts archived before the specified offset)
-          delete from participant_events_divulgence delete_events
+          -- Exercise events (consuming)
+          delete from participant_events_consuming_exercise delete_events
+          where
+            delete_events.event_offset <= $pruneUpToInclusive"""
+
+    val nonConsumingExercisesPruning =
+      SQL"""
+          -- Exercise events (non-consuming)
+          delete from participant_events_non_consuming_exercise delete_events
+          where
+            delete_events.event_offset <= $pruneUpToInclusive"""
+
+    val createsPruning =
+      SQL"""
+          -- Create events (only for contracts archived before the specified offset)
+          delete from participant_events_create delete_events
           where
             delete_events.event_offset <= $pruneUpToInclusive and
             exists (
               SELECT 1 FROM participant_events_consuming_exercise archive_events
               WHERE
                 archive_events.event_offset <= $pruneUpToInclusive AND
                 archive_events.contract_id = delete_events.contract_id
-            )""",
-      SQL"""
-          -- Create events (only for contracts archived before the specified offset)
-          delete from participant_events_create delete_events
+            )"""
+
+    val retroactiveDivulgencesPruning =
+      if (pruneAllDivulgedContracts)
+        SQL"""
+          -- Retroactive divulgence events
+          delete from participant_events_divulgence delete_events
+          where delete_events.event_offset <= $pruneUpToInclusive"""
+      else
+        SQL"""
+          -- Divulgence events (only for contracts archived before the specified offset)
+          delete from participant_events_divulgence delete_events
           where
             delete_events.event_offset <= $pruneUpToInclusive and
             exists (
               SELECT 1 FROM participant_events_consuming_exercise archive_events
               WHERE
                 archive_events.event_offset <= $pruneUpToInclusive AND
                 archive_events.contract_id = delete_events.contract_id
-            )""",
-      SQL"""
-          -- Exercise events (consuming)
-          delete from participant_events_consuming_exercise delete_events
-          where
-            delete_events.event_offset <= $pruneUpToInclusive""",
-      SQL"""
-          -- Exercise events (non-consuming)
-          delete from participant_events_non_consuming_exercise delete_events
-          where
-            delete_events.event_offset <= $pruneUpToInclusive""",
-    ).foreach(_.execute()(connection))
+            )"""
+
+    {
+      conditionalImmediateDivulgencePruning ++
```

Comment:
This query should come last to profit from the work already done by the `createsPruning`.

Reply:
👍

Reply:
Updated according to the discussion had offline.